Mathlib/Data/Matroid/IndepAxioms.lean

```diff
@@ -0,0 +1,565 @@
+independent set, then there exists `e ∈ J \ I` for which `insert e I` is independent.
+This is the axiom that appears in all most of the definitions.
+
+## Main definitions
+
+* `Matroid.matroidOfBase` constructs a matroid from the base axioms (using the definition of a
+  matroid).
+
+* `Matroid.matroidOfIndep` constructs a matroid described in terms of its independent sets
+  in full generality, using infinite versions of the axioms.
+
+* `Matroid.matroidOfIndepOfFinitary` constructs a `Finitary` matroid in terms of its independent
+  sets in the special case where independence of a set is determined only by that of its
+  finite subsets. This construction uses Zorn's lemma.
+
+* `Matroid.matroidOfIndepOfBdd` constructs a matroid in terms of its independence sets in
+  the case where there is some known absolute upper bound on the size of an independent set.
+  This uses the infinite version of the augmentation axiom, and gives a `FiniteRk` matroid.
+
+* `Matroid.matroidOfIndepOfBddAugment` is the same as the above, but with a finite augmentation
+  axiom.
+
+* `Matroid.matroidOfExistsFiniteBase` constructs a matroid from its bases, if it is known that one
+  of them is finite. This gives a `FiniteRk` matroid.
+
+* `Matroid.matroidOfBaseOfFinite` constructs a `Finite` matroid from its bases.
+
+* `Matroid.matroidOfIndepOfFinite` constructs a `Finite` matroid from its independent sets.
+
+* `Matroid.matroidOfIndepFinset` constructs a `Finitary` matroid on `α` from an independence
+  predicate on `Finset α`.
+
+* `Matroid.matroidOfIndepOfExistsMatroid` constructs a 'copy' of a matroid that is known only
+  existentially, but whose independence predicate is known explicitly.
+-/
+
+open Set
+
+variable {α : Type _}
+
+namespace Matroid
+
+/-- A constructor for matroids via the base axioms.
+  (In fact, just a wrapper for the definition of a matroid) -/
+def matroidOfBase (E : Set α) (Base : Set α → Prop) (exists_base : ∃ B, Base B)
+    (base_exchange : ExchangeProperty Base)
+    (maximality : ∀ X, X ⊆ E → ExistsMaximalSubsetProperty (∃ B, Base B ∧ · ⊆ B) X)
+    (support : ∀ B, Base B → B ⊆ E) : Matroid α :=
+  ⟨E, Base, exists_base, base_exchange, maximality, support⟩
+
+@[simp] theorem matroidOfBase_apply (E : Set α) (Base : Set α → Prop) (exists_base : ∃ B, Base B)
+    (base_exchange : ExchangeProperty Base)
+    (maximality : ∀ X, X ⊆ E → ExistsMaximalSubsetProperty (∃ B, Base B ∧ · ⊆ B) X)
+    (support : ∀ B, Base B → B ⊆ E) :
+    (matroidOfBase E Base exists_base base_exchange maximality support).Base = Base := rfl
+
+@[simp] theorem matroidOfBase_ground (E : Set α) (Base : Set α → Prop) (exists_base : ∃ B, Base B)
+    (base_exchange : ExchangeProperty Base)
+    (maximality : ∀ X, X ⊆ E → ExistsMaximalSubsetProperty (∃ B, Base B ∧ · ⊆ B) X)
+    (support : ∀ B, Base B → B ⊆ E) :
+    (matroidOfBase E Base exists_base base_exchange maximality support).E = E := rfl
```

Comment:
Does `@[simps]` on the `def` above generate these for you?

Reply:
In this one case, `@[simps]` does (although I've kept the `ofBase_apply` because it is sometimes useful to rewrite the predicate itself) but none of the other definitions of matroids below use the structure fields directly for the definition, so `@[simps]` seems not to work.

Reply:
Note that you can use `simps ground` or perhaps `simps E` to generate only the lemmas you think are useful.